telemetry/firmware/telemetry.c

```diff
@@ -0,0 +1,316 @@
+
+
+// CAN Mailboxes
+#define BRAKE_LIGHT_MBOX                   0
+#define BMS_CORE_MBOX                      1
+#define AIR_CONTROL_CRITICAL_MBOX          2
+#define AIR_CONTROL_SENSE_MBOX             3
+#define THROTTLE_MBOX                      4
+// handle another 15 mailboxes lol 
+
+
+// UART 
+
+// values within LINCR register
+#define ENABLE_LIN_AND_UART             LENA // can i leave this as LENA or does it have to be bit 3 (page 190)
+#define ENABLE_UART                     LCMD2
+#define ENABLE_RX                       LMCD1
+#define ENABLE_TX                       LCMD0
+
+#define UART_DATA                       LINDAT             // lin data register
+#define UART_BAUD                       9600
+
+
+// shutdown_error_flag codes within error var
+
+#define BSPD_FAIL                       0
+#define INERTIA_SWITCH_FAIL             1
+#define BOTS_FAIL                       2
+#define COCKPIT_ESTOP_FAIL              3
+#define HV_DISCONNECT_FAIL              4
+#define HVD_CONN_FAIL                   5
+#define PACK_CONN_FAIL                  6
+#define BMS_FAIL                        7
+#define IMD_FAIL                        8
+#define TSMS_FAIL                       9
+
+
+/*----- Global Variables -----*/
+
+volatile uint8_t gFlag = 0x00; // global flag
+
+volatile uint16_t shutdown_error_flag = 0x00; // system failure error code flag - default 0x00. 
+
+// CAN DATA to store & send out over UART 
+uint8_t accumulator_voltages[192] = {0}; // 192 wholeass bytes wtf
+uint8_t accumulator_temperatures[80] = {0}; 
+uint16_t throttle = 0x00;
+uint8_t state_of_charge = 0x00; 
+
+
+uint8_t can_recv_msg[8] = {0}; // CAN msg recieved
+
+
+
+uint8_t data_byte = 0x00; // for UART 
+
+
+
+typedef struct {
+    uint8_t ID; 
+    short data_length; 
+    uint8_t data[8]; 
+} CAN_MSG; 
+
+// QUEUES
+
+QUEUE(UART, uint8_t, UART_QUEUE_SIZE); 
+QUEUE(CAN, CAN_MSG *, CAN_QUEUE_SIZE); 
+
+voltatile struct queue_CAN CAN_queue; 
+voltatile struct queue_UART UART_queue; 
+
+
+
+// Define CAN struct --> to be created and set with ID & data 
+
+
+/*----- Interrupt(s) -----*/
+
+
+ISR(CAN_INT_vect){ // TODO: Write state machine for mailbox address switching
+/*----- Brake light mailbox -----*/
+    CANPAGE = (BRAKE_LIGHT_MBOX<<MOBNB0); // what does this do? 
+    if (bit_is_set(CANSTMOB, RXOK)) {
+		can_recv_msg[0] = CANMSG;   // PANIC PANIC
+		can_recv_msg[1] = CANMSG;   // brake analog voltage MSB       // subjectlines of the letter, have to read the letters in order
+		can_recv_msg[2] = CANMSG;   // brake analog voltage LSB
+		can_recv_msg[3] = CANMSG;   // is brake pressed? CF
+		can_recv_msg[4] = CANMSG;   // BSPD sense
+
+        // HUH ? wtf? below
+
+		can_recv_msg[5] = CANMSG;   // TSMS sense
+		can_recv_msg[6] = CANMSG;   // left e-stop sense
+		can_recv_msg[7] = CANMSG;   // GLVMS sense
+
+        /* For ALL CAN DATA, GENERATE CAN_MSG STRUCT AND PASS INTO CAN_QUEUE*/ 
+
+        CAN_MSG * msg = malloc(sizeof(CAN_MSG));  
```

Comment:
Don't need malloc. I would just do something like
```c
CAN_MSG* msg = {
    .ID = custom_ID,
    .length = 8,
    .data = can_recv_msg;
}
```

Reply:
Actually, come to think of it, you don't need a pointer here. You should just do
```c
CAN_MSG msg = {
    ...
}
```